syncr_backend/util/drop_util.py

```diff
@@ -0,0 +1,227 @@
+import os
+import shutil
+from typing import List
+from typing import Optional  # noqa
+from typing import Set
+from typing import Tuple
+
+from syncr_backend.constants import DEFAULT_DROP_METADATA_LOCATION
+from syncr_backend.constants import DEFAULT_FILE_METADATA_LOCATION
+from syncr_backend.init import drop_init
+from syncr_backend.init import node_init
+from syncr_backend.metadata import drop_metadata
+from syncr_backend.metadata.drop_metadata import DropMetadata
+from syncr_backend.metadata.drop_metadata import get_drop_location
+from syncr_backend.metadata.drop_metadata import save_drop_location
+from syncr_backend.metadata.file_metadata import FileMetadata
+from syncr_backend.network import send_requests
+from syncr_backend.util import crypto_util
+from syncr_backend.util import fileio_util
+
+
+def sync_drop(drop_id: bytes, save_dir: str):
+    """
+    Syncs a drop id
+    """
+    drop_peers = [("127.0.0.1", 1234)]
+    # TODO: get drop peers
+    raise Exception("Not Implemented")
+
+    add_drop_from_id(drop_id, save_dir)
+    drop_metadata = get_drop_metadata(drop_id, drop_peers)
+    for file_name, file_id in drop_metadata.files.items():
+        remaining_chunks = sync_drop_contents(
+            drop_id,
+            file_id,
+            drop_peers,
+            save_dir,
+        )
+        if not remaining_chunks:
+            fileio_util.mark_file_complete(file_name)
+
+
+class PermissionError(Exception):
+    pass
+
+
+def update_drop(drop_id: bytes) -> None:
+    """
+    Update a drop from a directory.
+
+    :param drop_id: The drop_id to update
+    :param peers: list of peers to get drop metadata from if missing
+
+    """
+    peers = [("127.0.0.1", 1234)]
+    # TODO: get drop peers
+    raise Exception("Not Implemented")
+
+    old_drop_metadata = get_drop_metadata(drop_id, peers)
+    priv_key = node_init.load_private_key_from_disk()
+    node_id = crypto_util.node_id_from_public_key(priv_key.public_key())
+
+    if old_drop_metadata.owner != node_id:
+        raise PermissionError("You are not the owner of this drop")
+
+    drop_directory = get_drop_location(drop_id)
+    (new_drop_m, new_files_m) = drop_init.make_drop_metadata(
+        path=drop_directory,
+        drop_name=old_drop_metadata.name,
+        owner=old_drop_metadata.owner,
+    )
+    new_drop_m.other_owners = old_drop_metadata.other_owners
+    new_drop_m.version = drop_metadata.DropVersion(
+        old_drop_metadata.version.version + 1,
+        crypto_util.random_int(),
+    )
+    # deletes the existing metadata files
```

Comment:
Don't delete the existing drop metadata files! We will need them.  Only delete old file metadata files.